File: carpeta/gestorAplicacion/adminVuelos/Aerolinea.java
// CLASE AEROLINEA 
package gestorAplicacion.adminVuelos;
import java.io.Serializable;
import java.util.ArrayList;
import gestorAplicacion.hangar.*;

// ALMACENA LA INFORMACION DE TODAS LAS AEROLINEAS CREADAS, ADEMAS DE LOS VUELOS Y AERONAVES ASOCIADOS A CADA UNA DE ELLAS 
// CON LOS METODOS NECESARIOS PARA ACCEDER A ESTA INFORMACION A TRAVES DE DISTINTOS PARAMETROS.
public class Aerolinea implements Serializable{

	private static final long serialVersionUID = 1L;
	
	//ATRIBUTOS
	private String nombre;
	private ArrayList<Aeronave> aeronaves = new ArrayList<Aeronave>();
	private ArrayList<Vuelo> vuelos = new ArrayList<Vuelo>();
	private static ArrayList<Aerolinea> aerolineas = new ArrayList<Aerolinea>();

	//CONSTRUCTOR
	public Aerolinea(String nombre) {
		this.nombre = nombre;
		aerolineas.add(this);
	}

	@Override
	public String toString() {
		return this.nombre;
	}

	//BUSCAR AEROLINEA
	
	// METODO DE CLASE QUE RECIBE UN NOMBRE DE AEROLINEA Y BUSCA ENTRE LAS AEROLINEAS DISPONIBLES SI HAY ALGUNA CON ESTE NOMBRE, 
	//SI ES ASÍ, RETORNA ESE OBJETO AEROLINEA, SI NO, DEVUELVE NULL.
	public static Aerolinea buscarAerolineaPorNombre(String nombre2)
	{
		Aerolinea retorno = null;
		for (int i = 0; i < Aerolinea.getAerolineas().size(); i++)
		{
			if( Aerolinea.getAerolineas().get(i).getNombre().equalsIgnoreCase(nombre2))
			{
				// SI ENCUENTRA UNA AEROLINEA CUYO NOMBRE COINCIDA CON EL PARAMETRO PASADO, IGNORANDO MINUSCULAS Y MAYUSCULAS,
				// RETORNA ESA AEROLINEA.
				retorno=  Aerolinea.getAerolineas().get(i);
			}
		}
		return retorno;
	}

	//BUSCAR VUELO POR...
	
	// RECIBE UNA LISTA DE VUELOS (ARRAYLIST<VUELO>) Y UN ID (INT) Y SE ENCARGA DE RETORNAR EL OBJETO VUELO QUE TENGA EL ID QUE PASAMOS 
	// COMO PARAMETRO, SI NO LO ENCUENTRA EN ESA LISTA DE VUELOS RETORNA NULL.
	public Vuelo buscarVueloPorID (ArrayList<Vuelo> vuelos, int ID)
	{
		for (int i = 0; i < vuelos.size(); i++)
		{
		  if (vuelos.get(i).getID() == ID )
		  {
			  return vuelos.get(i);
		  }
		}
		return null;
	}
	
	// RECIBE UNA LISTA DE VUELOS (ARRAYLIST<VUELO>) Y UN NOMBRE_AERONAVE(STRING) Y SE ENCARGA DE RETORNAR EL OBJETO VUELO QUE TENGA 
	// ASOCIADA LA AERONAVE QUE TIENE EL NOMBRE QUE PASAMOS COMO PARAMETRO, SI NO LO ENCUENTRA EN ESA LISTA DE VUELOS RETORNA NULL.
	public Vuelo buscarVueloPorAeronave (ArrayList<Vuelo> vuelos, String nombre_Aeronave)
	{
		for (int i = 0; i < vuelos.size(); i++)
		{
		  if (vuelos.get(i).getAeronave().getNombre().equals(nombre_Aeronave) )
		  {
			  return vuelos.get(i);
		  }
		}
		return null;
	}
	
	// RECIBE UNA LISTA DE VUELOS (ARRAYLIST<VUELO>) Y UN DESTINO(STRING) Y SE ENCARGA DE RETORNAR UNA LISTA DE VUELOS, 
	// CON TODOS LOS VUELOS QUE TENGAN ASOCIADO EL DESTINO QUE PASAMOS COMO PARAMETRO.
	public ArrayList<Vuelo> buscarVueloPorDestino (ArrayList<Vuelo> vuelos, String destino)
	{
		ArrayList<Vuelo> vuelosPorDestino = new ArrayList<Vuelo>();
		for (int i = 0; i < vuelos.size(); i++)
		{
		  if (vuelos.get(i).getDestino().equalsIgnoreCase(destino)) // cambio del equals
		  {
			  vuelosPorDestino.add(vuelos.get(i));
		  }
		}
		return vuelosPorDestino;
	}

	// RECIBE UNA LISTA DE VUELOS (ARRAYLIST<VUELO> ) Y UNA FECHA (STRING) Y SE ENCARGA DE RETORNAR UNA LISTA DE VUELOS, 
	// CON TODOS LOS VUELOS QUE TENGAN ASOCIADA LA FECHA QUE PASAMOS COMO PARAMETRO.
	public ArrayList<Vuelo> buscarVueloPorFecha (ArrayList<Vuelo> vuelos, String fecha)
	{
		ArrayList<Vuelo> vuelosPorFecha = new ArrayList<Vuelo>();
		for (int i = 0; i < vuelos.size(); i++)
		{
		  if (vuelos.get(i).getFecha_de_salida().equals(fecha))
		  {
			  vuelosPorFecha.add(vuelos.get(i));
		  }
		}
		return vuelosPorFecha;
	}

	//VUELOS DISPONIBLES
	
	// RECIBE UNA LISTA DE VUELOS (ARRAYLIST<VUELO>) Y SE ENCARGA DE RETORNAR UNA LISTA DE VUELOS, CON TODOS LOS VUELOS DE LA AEROLINEA
	// QUE NO ESTEN COMPLETOS, ES DECIR QUE SU ATRIBUTO ESTACOMPLETO SEA IGUAL A false.
	public ArrayList<Vuelo> vuelosDisponibles(ArrayList<Vuelo> vuelos)
	{
		ArrayList<Vuelo> vuelosDisponibles = new ArrayList<Vuelo>();
		for (int i = 0; i < vuelos.size(); i++)
		{
			if (!vuelos.get(i).isEstaCompleto())
			{
				vuelosDisponibles.add(vuelos.get(i));
			}
		}
		return vuelosDisponibles;
	}

	// AGREGAR O CANCELAR UN VUELO
	
	// RECIBE COMO PARAMETRO UN VUELO (VUELO) Y SE ENCARGA DE ANADIRLO A LA LISTA DE VUELOS DE LA AEROLINEA.
	public void agregarVuelo(Vuelo vuelo)
	{
		vuelos.add(vuelo);
	}

	// RECIBE COMO PARÁMETRO UN ENTERO, QUE CONTIENE EL ID DEL VUELO A ELIMINAR, Y SE ENCARGA DE RECORRER SU LISTA DE VUELOS PARA ELIMINAR 
	// EL VUELO QUE CONTENGA EL ID QUE COINCIDE CON EL PARAMETRO, SI LO ENCONTRO Y LO ELIMINO, RETORNA true, EN CASO CONTRARIO RETORNA false.
	public Boolean cancelarVuelo(int vuelo_a_eliminar)
	{
		for (int i = 0; i < vuelos.size(); i++)
		{
		  if (vuelos.get(i).getID() == vuelo_a_eliminar )
		  {
			  vuelos.remove(i);
			  return true;
		  }
		}
		return false;
	}

	// BUSCAR TIQUETE POR ID
	
	// METODO DE CLASE QUE RECIBE UN ID (INT) Y SE ENCARGA DE BUSCAR ENTRE CADA AEROLÍNEA Y ENTRE CADA VUELO DE ESTA AEROLINEA, 
	// SI HAY UN TIQUETE ASOCIADO EN LA LISTA DE TIQUETES DE CADA VUELO QUE TENGA POR ID EL QUE LE PASAMOS COMO PARÁMETRO Y RETORNAR ESTE TIQUETE, 
	// HACIENDO USO DEL METODO BUSCARTIQUETEPORID() DE VUELO.

	public static Tiquete BuscarTiquete(int ID)
	{
		ArrayList<Aerolinea> aerolineasDisponibles = Aerolinea.getAerolineas();
		for (int i = 0; i < aerolineasDisponibles.size(); i++)
		{
			Aerolinea aerolinea = aerolineasDisponibles.get(i);
			for (int j = 0; j < aerolinea.getVuelos().size(); j++)
			{

				Vuelo vuelo = aerolinea.getVuelos().get(j);
				Tiquete tiquete_buscado = vuelo.buscarTiquetePorID(vuelo.getTiquetes(), ID);
				if (tiquete_buscado != null)
				{
					return tiquete_buscado;
				}
			}
		}
		return null;
	}
	  
	// SETTERS Y GETTERS

	public String getNombre() {
		return nombre;
	}


	public void setNombre(String nombre) {
		this.nombre = nombre;
	}


	public ArrayList<Vuelo> getVuelos() {
		return vuelos;
	}


	public void setVuelos(ArrayList<Vuelo> vuelos) {
		this.vuelos = vuelos;
	}

	public ArrayList<Aeronave> getAeronaves() {
		return aeronaves;
	}

	public void setAviones(ArrayList<Aeronave> aviones) {
		this.aeronaves = aviones;
	}

	public static ArrayList<Aerolinea> getAerolineas() {
		return aerolineas;
	}

	public static void setAerolineas(ArrayList<Aerolinea> aerolineas) {
		Aerolinea.aerolineas = aerolineas;
	}

}

File: carpeta/gestorAplicacion/adminVuelos/Aerolinea.py
from excepciones.ErrorAsignacion import ExcepcionIdTiquete
from gestorAplicacion.hangar import *
from gestorAplicacion.adminVuelos.Vuelo import Vuelo
from gestorAplicacion.adminVuelos.Tiquete import Tiquete

# ALMACENA LA INFORMACION DE TODAS LAS AEROLINEAS CREADAS, ADEMAS DE LOS VUELOS Y AERONAVES ASOCIADOS A CADA UNA DE ELLAS 
# CON LOS METODOS NECESARIOS PARA ACCEDER A ESTA INFORMACION A TRAVES DE DISTINTOS PARAMETROS.
class Aerolinea():

    #ATRIBUTOS
    _aerolineas = []

    #CONSTRUCTOR
    def __init__(self, nombre):
        self._aeronaves = []
        self._vuelos = []

        self._nombre = nombre
        Aerolinea._aerolineas.append(self)

    def toString(self):
        return self._nombre

    #BUSCAR AEROLINEA

    # METODO DE CLASE QUE RECIBE UN NOMBRE DE AEROLINEA Y BUSCA ENTRE LAS AEROLINEAS DISPONIBLES SI HAY ALGUNA CON ESTE NOMBRE, 
    #SI ES ASÍ, RETORNA ESE OBJETO AEROLINEA, SI NO, DEVUELVE NULL.
    @staticmethod
    def buscarAerolineaPorNombre(nombre2):
        retorno = None
        i = 0
        while i < len(Aerolinea.getAerolineas()):
            if Aerolinea.getAerolineas()[i].getNombre().lower() == nombre2.lower():
                # SI ENCUENTRA UNA AEROLINEA CUYO NOMBRE COINCIDA CON EL PARAMETRO PASADO, IGNORANDO MINUSCULAS Y MAYUSCULAS,
                # RETORNA ESA AEROLINEA.
                retorno = Aerolinea.getAerolineas()[i]
            i += 1
        return retorno

    #BUSCAR VUELO POR...

    # RECIBE UNA LISTA DE VUELOS (ARRAYLIST<VUELO>) Y UN ID (INT) Y SE ENCARGA DE RETORNAR EL OBJETO VUELO QUE TENGA EL ID QUE PASAMOS 
    # COMO PARAMETRO, SI NO LO ENCUENTRA EN ESA LISTA DE VUELOS RETORNA NULL.
    def buscarVueloPorID(self, vuelos, ID):
        i = 0
        while i < len(vuelos):
            if vuelos[i].getID() == ID:
                return vuelos[i]
            i += 1
        return None

    # RECIBE UNA LISTA DE VUELOS (ARRAYLIST<VUELO>) Y UN NOMBRE_AERONAVE(STRING) Y SE ENCARGA DE RETORNAR EL OBJETO VUELO QUE TENGA 
    # ASOCIADA LA AERONAVE QUE TIENE EL NOMBRE QUE PASAMOS COMO PARAMETRO, SI NO LO ENCUENTRA EN ESA LISTA DE VUELOS RETORNA NULL.
    def buscarVueloPorAeronave(self, vuelos, nombre_Aeronave):
        i = 0
        while i < len(vuelos):
            if vuelos[i].getAeronave().getNombre().casefold() ==  nombre_Aeronave.casefold():
                return vuelos[i]
            i += 1
        return None

    # RECIBE UNA LISTA DE VUELOS (ARRAYLIST<VUELO>) Y UN DESTINO(STRING) Y SE ENCARGA DE RETORNAR UNA LISTA DE VUELOS, 
    # CON TODOS LOS VUELOS QUE TENGAN ASOCIADO EL DESTINO QUE PASAMOS COMO PARAMETRO.
    def buscarVueloPorDestino(self, vuelos, destino):
        vuelosPorDestino = []
        i = 0
        while i < len(vuelos):
            if vuelos[i].getDestino().casefold() == destino.casefold(): # cambio del equals
                vuelosPorDestino.append(vuelos[i])
            i += 1
        return vuelosPorDestino

    # RECIBE UNA LISTA DE VUELOS (ARRAYLIST<VUELO> ) Y UNA FECHA (STRING) Y SE ENCARGA DE RETORNAR UNA LISTA DE VUELOS, 
    # CON TODOS LOS VUELOS QUE TENGAN ASOCIADA LA FECHA QUE PASAMOS COMO PARAMETRO.
    def buscarVueloPorFecha(self, vuelos, fecha):
        vuelosPorFecha = []
        i = 0
        while i < len(vuelos):
            if vuelos[i].getFecha_de_salida() == fecha:
                vuelosPorFecha.append(vuelos[i])
            i += 1
        return vuelosPorFecha

    #VUELOS DISPONIBLES

    # RECIBE UNA LISTA DE VUELOS (ARRAYLIST<VUELO>) Y SE ENCARGA DE RETORNAR UNA LISTA DE VUELOS, CON TODOS LOS VUELOS DE LA AEROLINEA
    # QUE NO ESTEN COMPLETOS, ES DECIR QUE SU ATRIBUTO ESTACOMPLETO SEA IGUAL A false.
    def vuelosDisponibles(self, vuelos):
        vuelosDisponibles = []
        i = 0
        while i < len(vuelos):
            if not vuelos[i].isEstaCompleto():
                vuelosDisponibles.append(vuelos[i])
            i += 1
        return vuelosDisponibles

    # AGREGAR O CANCELAR UN VUELO

    # RECIBE COMO PARAMETRO UN VUELO (VUELO) Y SE ENCARGA DE ANADIRLO A LA LISTA DE VUELOS DE LA AEROLINEA.
    def agregarVuelo(self, vuelo):
        self._vuelos.append(vuelo)

    # RECIBE COMO PARÁMETRO UN ENTERO, QUE CONTIENE EL ID DEL VUELO A ELIMINAR, Y SE ENCARGA DE RECORRER SU LISTA DE VUELOS PARA ELIMINAR 
    # EL VUELO QUE CONTENGA EL ID QUE COINCIDE CON EL PARAMETRO, SI LO ENCONTRO Y LO ELIMINO, RETORNA true, EN CASO CONTRARIO RETORNA false.
    def cancelarVuelo(self, vuelo_a_eliminar):
        i = 0
        while i < len(self._vuelos):
            if self._vuelos[i].getID() == vuelo_a_eliminar:
                self._vuelos.pop(i)
                return True
            i += 1
        return False

    # BUSCAR TIQUETE POR ID

    # METODO DE CLASE QUE RECIBE UN ID (INT) Y SE ENCARGA DE BUSCAR ENTRE CADA AEROLÍNEA Y ENTRE CADA VUELO DE ESTA AEROLINEA, 
    # SI HAY UN TIQUETE ASOCIADO EN LA LISTA DE TIQUETES DE CADA VUELO QUE TENGA POR ID EL QUE LE PASAMOS COMO PARÁMETRO Y RETORNAR ESTE TIQUETE, 
    # HACIENDO USO DEL METODO BUSCARTIQUETEPORID() DE VUELO.

    @staticmethod
    def BuscarTiquete(ID):
        tiquete_buscado = None
        aerolineasDisponibles = Aerolinea.getAerolineas()
        i = 0
        while i < len(aerolineasDisponibles):
            aerolinea = aerolineasDisponibles[i]
            j = 0
            while j < len(aerolinea.getVuelos()):

                vuelo = aerolinea.getVuelos()[j]
                tiquete_buscado = vuelo.buscarTiquetePorID(vuelo.getTiquetes(), ID)
                if tiquete_buscado is not None:
                    return tiquete_buscado
                j += 1
            i += 1
        
        if tiquete_buscado == None:
            raise ExcepcionIdTiquete(ID)
        return None

    # SETTERS Y GETTERS

    def getNombre(self):
        return self._nombre


    def setNombre(self, nombre):
        self._nombre = nombre


    def getVuelos(self):
        return self._vuelos


    def setVuelos(self, vuelos):
        self._vuelos = vuelos

    def getAeronaves(self):
        return self._aeronaves

    def setAviones(self, aviones):
        self._aeronaves = aviones

    @staticmethod
    def getAerolineas():
        return Aerolinea._aerolineas

    @staticmethod
    def setAerolineas(aerolineas):
        Aerolinea._aerolineas = aerolineas
